Equipment persistence

raicovx opened the pull request:
Addressing #59 

- Saves per checkpoint (in order to respawn with equipment, will need tweaking for survival mode/preference).
- Save command to export equipment to file
- Load command to import equipment from file
- Clear command to clear currently persisted equipment and reset players to default equipment
- Added ConVar to disable equipment persistence, on by default

Alienmario reviewed (changes requested):
Good first PR! Added comments where it needs some tweaking.

Please also adjust the code style - mostly { on new line and spaces after keywords "if (".
The save spots might need some adjustments. I have yet to test it.

I'd probably move more logic to a new include like manager_persistence.inc. Don't like having too much logic in the main file. What do you think?

raicovx commented:
Thanks for the review, I'm very used to garbage collection haha.

Originally I had a separate file but I ran into issues with circular includes that I want too sure how to deal with. The checkpoint functions needed to use the equipment manager and the equipment manager used the coopequipment. Id imagine I could get away with being less strict with the type on the EM side.

I'll fix these issues up tomorrow :)

raicovx commented on scripting/include/srccoop/checkpoint.inc:
```diff
@@ -290,24 +316,111 @@ enum struct CCoopSpawn
+		{
+			pPlayer.SetHealth(equipment.m_iHealth);
+		}
+		else if(this.m_pDefaultEquipment.m_iHealth > 0)
```
It should do, I originally had it as a fail-safe for invalid values being saved. When survival mode is implemented we may need to reconsider how the save is handled.

raicovx commented on scripting/include/srccoop/checkpoint.inc:
```diff
@@ -149,6 +149,32 @@ enum struct CCoopEquipment
+        this.m_smEquipmentMap = new StringMap();
+    }
+
+    void Clear()
```
This one I'm not entirely sure what to do with. If I clear the handles in the store function I get invalid handle errors when I try to get the array later. I've implemented it for the clear function though. 

Which also raises another issue, I was trying to implement clear so that it would release the memory when a client disconnects via OnClientDisconnect but this was running in between map changes as well which broke the whole thing. Any ideas for how to handle client disconnection without deleting the equipment on map change?

raicovx replied:
This should be resolved now. I hooked into player_disconnect event to dealloc the weapon list when a player disconnects. this event doesn't fire on map change and should handle all other disconnect events (such as timeout) from my research

raicovx commented on scripting/include/srccoop/checkpoint.inc:
```diff
@@ -290,24 +316,111 @@ enum struct CCoopSpawn
-		if(this.m_pDefaultEquipment.m_iHealth > 0)
+		float flStartTime = GetEngineTime();
+		CCoopEquipment existingEquipment;
+		existingEquipment.Initialize();
```
This has been addressed

raicovx commented on scripting/include/srccoop/checkpoint.inc:
```diff
@@ -290,24 +316,111 @@ enum struct CCoopSpawn
+	void SavePlayerEquipment()
+	{
+		float flStartTime = GetEngineTime();
+		for(int iClient = 0; iClient < MAXPLAYERS + 1; iClient++)
```
I've changed all instances of MAXPLAYERS i could find

raicovx commented on scripting/include/srccoop/checkpoint.inc:
```diff
@@ -290,24 +316,111 @@ enum struct CCoopSpawn
+				continue;
+
+			CBasePlayer pPlayer = CBasePlayer(iClient);
+			if(!pPlayer.IsValid() || !pPlayer.IsInGame())
```
Updated all instances of this i could find

raicovx commented on scripting/srccoop.sp:
```diff
@@ -273,11 +281,13 @@ public void OnClientPutInServer(int client)
 	DHookEntity(hkChangeTeam, false, client, _, Hook_PlayerChangeTeam);
 	DHookEntity(hkShouldCollide, false, client, _, Hook_PlayerShouldCollide);
 	GreetPlayer(client);
+	SetSteamId(client);
```
Refactor into OnClientAuthorized, i stuck with the format i was using previously due to the dynamic size of char[] auth

raicovx commented on scripting/srccoop.sp:
```diff
@@ -745,3 +770,209 @@ public Action Command_DumpMapEntities(int iArgs)
+void GetSaveDataPath(char buffer[PLATFORM_MAX_PATH], char szFileName[64]){
+	char szFolder[10] = "savedata/";
+
+	char relativePath[PLATFORM_MAX_PATH + 64];
```
That makes sense, updated

raicovx commented on scripting/srccoop.sp:
```diff
@@ -745,3 +770,209 @@ public Action Command_DumpMapEntities(int iArgs)
 }
+
+void GetSaveDataPath(char buffer[PLATFORM_MAX_PATH], char szFileName[64]){
+	char szFolder[10] = "savedata/";
```
Agreed, updated

raicovx commented on scripting/srccoop.sp:
```diff
@@ -745,3 +770,209 @@ public Action Command_DumpMapEntities(int iArgs)
+	strcopy(relativePath, sizeof(relativePath), szFolder);
+	StrCat(relativePath, sizeof(relativePath), szFileName);
+
+	BuildPath(Path_SM, buffer, sizeof(buffer), relativePath);
```
Thanks for the tip, updated.

raicovx commented on scripting/srccoop.sp:
```diff
@@ -745,3 +770,209 @@ public Action Command_DumpMapEntities(int iArgs)
+
+	KeyValues kv = new KeyValues("SaveData");
+	bool shouldSave = false;
+	for(int i = 0; i < MAXPLAYERS + 1; i++)
```
as mentioned above, updated

raicovx commented on scripting/srccoop.sp:
```diff
@@ -745,3 +770,209 @@ public Action Command_DumpMapEntities(int iArgs)
+			continue;
+		
+		CCoopEquipment equipment; 
+		equipment.Initialize();
```
Removed

raicovx commented on scripting/srccoop.sp:
```diff
@@ -745,3 +770,209 @@ public Action Command_DumpMapEntities(int iArgs)
+		return Plugin_Handled;	
+	}
+
+	for(int i = 0; i < MAXPLAYERS + 1; i++)
```
Now uses MaxClients

raicovx commented on scripting/srccoop.sp:
```diff
@@ -745,3 +770,209 @@ public Action Command_DumpMapEntities(int iArgs)
+			continue;
+
+		CBasePlayer pPlayer = CBasePlayer(i);
+		if(!pPlayer.IsValid() || !pPlayer.IsInGame())
```
Removed IsInGame

raicovx commented on scripting/srccoop.sp:
```diff
@@ -745,3 +770,209 @@ public Action Command_DumpMapEntities(int iArgs)
+
+	g_pEquipmentManager.Clear();
+
+	for(int i = 0; i < MAXPLAYERS + 1; i++)
```
Now uses MaxClients

raicovx commented on scripting/srccoop.sp:
```diff
@@ -745,3 +770,209 @@ public Action Command_DumpMapEntities(int iArgs)
+			continue;
+
+		CBasePlayer pPlayer = CBasePlayer(i);
+		if(!pPlayer.IsValid() || !pPlayer.IsInGame())
```
Removed IsInGame

raicovx commented on scripting/srccoop.sp:
```diff
@@ -745,3 +770,209 @@ public Action Command_DumpMapEntities(int iArgs)
+	if(!kv.ImportFromFile(szFullPath))
+	{
+		MsgReply(iClient, "Failed to parse KeyValues file %s", szFullPath);
+		return Plugin_Handled;	
```
Added delete kv

raicovx commented on scripting/srccoop.sp:
```diff
@@ -745,3 +770,209 @@ public Action Command_DumpMapEntities(int iArgs)
+		if(!kv.JumpToKey("weapons"))
+		{
+			MsgReply(iClient, "Unable to find weapons section in KeyValues file");
+			return Plugin_Handled;	
```
Added delete kv and CloseHandle for weapons list

raicovx commented on scripting/srccoop.sp:
```diff
@@ -745,3 +770,209 @@ public Action Command_DumpMapEntities(int iArgs)
+		if(!kv.JumpToKey("ammo"))
+		{
+			MsgReply(iClient, "Unable to find ammo section in KeyValues file");
+			return Plugin_Handled;	
```
Added delete kv and CloseHandle for weapons list

raicovx commented on scripting/srccoop.sp:
```diff
@@ -745,3 +770,209 @@ public Action Command_DumpMapEntities(int iArgs)
+		g_SpawnSystem.StripPlayer(pPlayer);
+		g_SpawnSystem.SpawnPlayerEquipment(pPlayer);
+
+		PrintToServer("Loaded Equipment for player %s", g_szSteamIds[i]);
```
Updated to use MsgReply

raicovx commented on scripting/srccoop.sp:
```diff
@@ -682,6 +692,21 @@ void GreetPlayer(int client)
+	}
+}
+
+void ClearSteamId(int client){
```
Removed

raicovx commented:
I've added the ability to load equipment from file for an individual player: sc_load <FileName> [<PlayerName>]

Alienmario commented:
Merged latest codebase that was previously in the survival branch. This means you'll need a newer Sourcemod as per the readme. As a bonus, having most logic in a separate include should be now possible without circular dependency issues, as shown in manager_survival.inc.
I'll try to do the review soon.

Alienmario reviewed (commented):
- The server prints are too verbose. Either remove them or turn them into LogDebug.
- `for (int iClient = 1; iClient < MaxClients; iClient++)` Needs to be <= (last client will be equal to MaxClients)

raicovx commented on scripting/include/srccoop/checkpoint.inc:
```diff
@@ -155,6 +155,44 @@ enum struct CCoopEquipment
+    void Initialize()
+    {
+        this.m_smEquipmentMap = new StringMap();
+    }
```
This should be fixed. Didn't get any warnings with compiler.

raicovx commented on scripting/include/srccoop/checkpoint.inc:
```diff
@@ -155,6 +155,44 @@ enum struct CCoopEquipment
+
+    void StoreEquipment(char[] steamId, CCoopEquipment equipment)
+	{
+		this.m_smEquipmentMap.SetArray(steamId, equipment, sizeof(equipment), true);
```
I now check for existing equipment, close it and remove the key before adding a new key / array with replace set to false

raicovx commented on scripting/include/srccoop/checkpoint.inc:
```diff
@@ -155,6 +155,44 @@ enum struct CCoopEquipment
+
+		PrintToServer("[SRCCOOP] Clearing equipment for %s", steamId);
+
+		CloseHandle(equipment.m_pWeaponList);
```
Very true. I've added this and replaced instances of CloseHandle

raicovx commented:
I think i've resolved your comments, i also removed the print to servers (they were more for my testing tbh) and fixed instances where I used < rather than <=.

I'll have a look at extracting the logic into the separate include next

raicovx commented:
I've extracted the logic into manager_persistence. I also made some improvements to the equipment persistence. Primary ammo is now stored and loaded to/from player equipment. I had included secondary ammo, but that just appears to use the ammo in the ammo collection.

raicovx commented:
All looking good from my testing now. Health, Armor, Clip and Ammo reserve maintained between maps. Saving and loading also working

raicovx commented:
Brining in another person revealed some issues, also need to deal with dead people on map change/checkpoint

raicovx commented:
Fixes included in [fbb7430] 
(https://github.com/ampreeT/SourceCoop/pull/71/commits/fbb74300ae8e276826dedb921105ac639528e357)
- Fix for saving multiple players, omitted the code to return to parent node on last change.
- Equipment cleared for dead players at the end of the map. Avoids saving empty equipment and they then get the map default.
- Equipment save disabled at checkpoints for map restart survival mode, allowing players to start again with what they had at map change.

Alienmario commented:
Sorry for the delay.

- Is saving at each checkpoint required for something or could it be made an optional feature (to save on map end only)?
- Should dead players be ignored in SavePlayerEquipment?

Alienmario commented:
Going to merge & release very soon, presuming there are no last comments.

Balimbanana commented:
Do you think it should also store transition position/angles offset from the landmark? On transition, you could get the landmark entity's position, then get all living players positions - landmark position (if inside `trigger_transition` *or* `trigger_changelevel` bounds) and restore on the next map as an offset of the stored named landmark.
You would also need to check if a checkpoint/spawnpoint has been set and cancel any position restores in case it prevents the player from catching up from slow loading or something.

I could write some of it up, depending on if we want to ExplodeString on the string array of equipment to get offset position and angles. Would also need to store `IsDucked()` and restore on the next map as `StartDucking()` if true. That will allow transitioning through vents and other debris on some maps.

There are a couple adjustments that would have to be made to the EDT's on transition such as the transition to `bm_c4a1b` to fire `B_Exit_door,Open,,1,-1` and the transition to `bm_c4a3d1` to fire `elevator_exit_door1,Open,,1,-1` along with `elevator_exit_door2` because players will be transitioned in to a closed room or elevator on those map changes.
There will most likely be other transitions that would need to be addressed, but it makes for a much smoother experience, and almost all scenarios will work as intended.

Alienmario commented:
It could work, if that's something you want to dedicate time into. Not sure where you'd use ExplodeString, though I'd prefer all organized within enum structs.


This might be a longer endeavor so I'm going to merge this PR first.

Alienmario commented:
@raicovx I forgot to ask, would you like to be featured in credits in the readme?

raicovx commented:
Sorry for the delay, looks like you've sorted it.

Sure, a credit would be awesome. Thanks 🙂
